pkg/reconciler/v1alpha1/autoscaling/kpa_scaler.go

```diff
@@ -129,44 +129,44 @@ func (ks *kpaScaler) Scale(ctx context.Context, kpa *kpa.PodAutoscaler, desiredS
 	}
 	currentScale := scl.Spec.Replicas
 
-	if desiredScale == 0 {
-		// We should only scale to zero when both of the following conditions are true:
-		//   a) The KPA has been active for atleast the stable window, after which it gets marked inactive
-		//   b) The KPA has been inactive for atleast the grace period
-
-		config := ks.getAutoscalerConfig()
-
-		if kpa.Status.IsActivating() { // Active=Unknown
-			// Don't scale-to-zero during activation
-			desiredScale = ScaleUnknown
-		} else if kpa.Status.IsReady() { // Active=True
-			// Don't scale-to-zero if the KPA is active
-
-			// Only let a revision be scaled to 0 if it's been active for at
-			// least the stable window's time.
-			if kpa.Status.CanMarkInactive(config.StableWindow) {
-				return desiredScale, nil
-			}
-			// Otherwise, scale down to 1 until the idle period elapses
-			desiredScale = 1
-		} else { // Active=False
-			// Don't scale-to-zero if the grace period hasn't elapsed
-			if !kpa.Status.CanScaleToZero(config.ScaleToZeroGracePeriod) {
-				return desiredScale, nil
-			}
-		}
```

Comment:
By dropping this we drop two very important guards for error free operation:

1. We shall not scale a revision down before it was actually deployed correctly.
2. We shall not kill all pods before the route has not been successfully switched to the activator

How do you envision to keep these premises alive?

Reply:
Thanks for reviewing my wip pr. but i am not quit sure about your concerns. 
For the point 1, Do you mean scale down to zero or just scale down to no-zero but smaller replica? I only find the guard to not scale to 0 before the revision actually deployed correctly, but didn't find guard not scale down to the smaller but no-zero replica. I can find that we make sure the at least 1 replica.
For the point2, it focus on leaving time for istio to update it's route. Both for switching to activator or to the revison, we just specific a approximate time. 
I think we can keeping record the mock stat once all the stats in stable window is empty for about several times. and we can set the kpa.status to inactive at the last time of record the mock stat.

Reply:
Towards 1.: Yes I meant scale-to-zero

Towards 2.: We need to deactivate the revision, leave some time (30s minimum currently) and only then scale the pods of the revision down to 0. Deactivating it will cause the route to flip over to the activator, which doesn't happen instantly. How is that achieved in this PR? (Sorry if I missed it)

Reply:
Hi, sorry for late reply, I am still trying on this.

Reply:
@lvjing2 I'm not sure the approach you're taking here is going to make things simpler to understand ultimately. Is simplification the goal of this PR?

Also note that a fake metric can not be used for point 1 I raised above. I'm not sure how to achieve that guard without communication of the Revision's state?

Reply:
The initial goal is to simply the autoscaler, but then I start try to decouple the status of kpa/revision with autoscaler. If we can decouple them, then we needn't to check so many status of kpa. 
In the point 1, cause there is always a mock stat in the stable window, then it can't scale down to 0 from the beginning to the end of `keepAliveTimes * stableWindow`. In this situation, we can't get the exact time when everything is deployed correctly,  so I think we need set keepAliveTimes=3 at least.